src/uct/ib/mlx5/gdaki/gdaki.cuh

```diff
@@ -0,0 +1,354 @@
+
+    cseg.opmod_idx_opcode = doca_gpu_dev_verbs_bswap32(
+            ((uint32_t)wqe_idx << DOCA_GPUNETIO_VERBS_WQE_IDX_SHIFT) | opcode);
+    cseg.qpn_ds           = doca_gpu_dev_verbs_bswap32(qp->sq_num_shift8 | ds);
+    cseg.fm_ce_se         = ctrl_flags;
+
+    rseg.raddr = doca_gpu_dev_verbs_bswap64(raddr);
+    rseg.rkey  = rkey;
+
+    dseg.byte_count = doca_gpu_dev_verbs_bswap32(bytes);
+    dseg.lkey       = lkey;
+    dseg.addr       = doca_gpu_dev_verbs_bswap64(laddr);
+
+    if (is_atomic) {
+        atseg.swap_add = doca_gpu_dev_verbs_bswap64(add);
+        doca_gpu_dev_verbs_store_wqe_seg(atseg_ptr, (uint64_t*)&(atseg));
+    }
+
+    doca_gpu_dev_verbs_store_wqe_seg(cseg_ptr, (uint64_t*)&(cseg));
+    doca_gpu_dev_verbs_store_wqe_seg(rseg_ptr, (uint64_t*)&(rseg));
+    doca_gpu_dev_verbs_store_wqe_seg(dseg_ptr, (uint64_t*)&(dseg));
+}
+
+UCS_F_DEVICE void uct_rc_mlx5_gda_db(struct doca_gpu_dev_verbs_qp *qp,
+                                     uint64_t wqe_base, unsigned count,
+                                     uint64_t flags)
+{
+    cuda::atomic_ref<uint64_t, cuda::thread_scope_device> ref(
+            qp->sq_ready_index);
+    uint64_t wqe_base_orig = wqe_base;
+    while (!ref.compare_exchange_strong(wqe_base, wqe_base + count,
+                                        cuda::std::memory_order_relaxed)) {
+        wqe_base = wqe_base_orig;
+    }
+
+    if (!(flags & UCT_DEVICE_FLAG_NODELAY)) {
+        return;
+    }
+
+    doca_gpu_dev_verbs_lock<DOCA_GPUNETIO_VERBS_RESOURCE_SHARING_MODE_GPU>(
+            &qp->sq_lock);
+    doca_gpu_dev_verbs_ring_db<DOCA_GPUNETIO_VERBS_SYNC_SCOPE_GPU,
+                               DOCA_GPUNETIO_VERBS_GPU_CODE_OPT_DEFAULT>(
+            qp, qp->sq_ready_index);
+    doca_priv_gpu_dev_verbs_update_dbr<DOCA_GPUNETIO_VERBS_QP_SQ>(
+            qp, qp->sq_ready_index);
+    doca_gpu_dev_verbs_ring_db<DOCA_GPUNETIO_VERBS_SYNC_SCOPE_GPU,
+                               DOCA_GPUNETIO_VERBS_GPU_CODE_OPT_DEFAULT>(
+            qp, qp->sq_ready_index);
+    doca_gpu_dev_verbs_unlock<DOCA_GPUNETIO_VERBS_RESOURCE_SHARING_MODE_GPU>(
+            &qp->sq_lock);
+}
+
+template<uct_device_level_t level>
+UCS_F_DEVICE ucs_status_t uct_rc_mlx5_gda_ep_put_single(
+        uct_device_ep_h tl_ep, const uct_device_mem_element_t *tl_mem_elem,
+        const void *address, uint64_t remote_address, size_t length,
+        uint64_t flags, uct_device_completion_t *comp)
+{
+    uct_rc_gdaki_dev_ep_t *ep = (uct_rc_gdaki_dev_ep_t*)tl_ep;
+    uct_rc_gdaki_device_mem_element_t *mem_elem =
+            (uct_rc_gdaki_device_mem_element_t*)tl_mem_elem;
+    struct doca_gpu_dev_verbs_qp *qp = ep->qp;
+    enum doca_gpu_dev_verbs_wqe_ctrl_flags cflag =
+            (doca_gpu_dev_verbs_wqe_ctrl_flags)0;
+    uint64_t wqe_idx;
+    unsigned lane_id;
+    unsigned num_lanes;
+    uint32_t fc;
+
+    uct_rc_mlx5_gda_exec_init<level>(&lane_id, &num_lanes);
+    uct_rc_mlx5_gda_reserv_wqe<level>(qp, 1, lane_id, wqe_idx);
```

Comment:
```C++
template<uct_device_level_t level>
UCS_F_DEVICE void
uct_rc_mlx5_gda_reserv_wqe(uint64_t &wqe_base)
{
    if (level == UCT_DEVICE_LEVEL_WARP) {
        wqe_base = __shfl_sync(0xffffffff, wqe_base, 0);
    } else if (level == UCT_DEVICE_LEVEL_BLOCK) {
        __syncthreads();
    }
}

...
    if (lane_id == 0) {
        wqe_base = uct_rc_mlx5_gda_reserv_wqe_thread(qp, count);
    }
    uct_rc_mlx5_gda_reserv_wqe<level>(wqe_idx);
```

Because only the second part of `uct_rc_mlx5_gda_reserv_wqe` depends on template parameter.